### PR TITLE
feat: initial code for plugin architecture

### PR DESCRIPTION
**Pull request by igpetrov**
## Description

feat: initial code for plugin architecture

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

### CODE REVIEW

File: operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java
Comment (npepinpe) [inline operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java:169]: :wrench: I would recommend re-throwing it as an `UncheckedIOException`, so the type still carries meaning.

File: operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java
Comment (npepinpe) [inline operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java:223]: :+1: Nice catch. I would add this information on the `ExternalJarClassLoader` java doc - we could even overload the `close` implementation of that class and log a warning there, e.g. `Closing the external JAR class loader may lead to ClassNotFoundException; only close if it's truly necessary`.

File: tasklist/common/src/main/java/io/camunda/tasklist/es/ElasticsearchConnector.java
Comment (npepinpe) [inline tasklist/common/src/main/java/io/camunda/tasklist/es/ElasticsearchConnector.java:220]: :wrench: I think this is because `ThreadContextUtil.callWithClassLoader` throws an exception in its signature, so you're forced to wrap it here, even though your plugin does not throw. How about we add a new `ThreadContextUtil.supplyWithClassLoader` which takes a `Supplier<V>` instead of a `Callable<V>`? I'm sure there would be other usages elsewhere too.

File: tasklist/common/src/main/java/io/camunda/tasklist/es/ElasticsearchConnector.java
Comment (npepinpe) [inline tasklist/common/src/main/java/io/camunda/tasklist/es/ElasticsearchConnector.java:223]: :x: As a user, I would have no idea what to do here. Why did it fail to load? What should I change? Please add a descriptive error message.

File: tasklist/common/src/main/java/io/camunda/tasklist/property/InterceptorPluginProperties.java
Comment (npepinpe) [inline tasklist/common/src/main/java/io/camunda/tasklist/property/InterceptorPluginProperties.java:60]: :x: For the equals and hashcode, here and in the other config classes, please refer to [the code style here](https://github.com/camunda/camunda/wiki/Code-Style#equalshashcode-implementations). I don't think we need to use common langs for such small things.

File: zeebe/exporters/elasticsearch-exporter/pom.xml
Comment (npepinpe) [inline zeebe/exporters/elasticsearch-exporter/pom.xml:92]: :x: Generally avoid adding dependencies to a module (even if they're technically used elsewhere) if it's only for small things (like equals/hashcode implementations). Every time this means coupling us more to this dependency, so we should evaluate if it's worth it (and in this case, probably not). The same for the other modules were this was added due to the hashcode/equals implementations (and please also change those).

File: zeebe/exporters/elasticsearch-exporter/src/main/java/io/camunda/zeebe/exporter/RestClientFactory.java
Comment (npepinpe) [inline zeebe/exporters/elasticsearch-exporter/src/main/java/io/camunda/zeebe/exporter/RestClientFactory.java:110]: :x: Same comments as in the other modules/connectors (i.e. error handling, supplyWithClassLoader, etc.)

File: tasklist/common/src/main/java/io/camunda/tasklist/os/OpenSearchConnector.java
Comment (npepinpe) [inline tasklist/common/src/main/java/io/camunda/tasklist/os/OpenSearchConnector.java:408]: :x: Same comment as in the ES connector.

File: zeebe/exporters/opensearch-exporter/src/main/java/io/camunda/zeebe/exporter/opensearch/RestClientFactory.java
Comment (npepinpe) [inline zeebe/exporters/opensearch-exporter/src/main/java/io/camunda/zeebe/exporter/opensearch/RestClientFactory.java:155]: :x: Same comments as with the ES exporter/connectors/etc.

**Review by npepinpe — DISMISSED** [review]
:rocket: Thanks Igor! Good catch with the class loader close method. It makes sense now that we know - after all, it has to open the JAR file to look for them, so if we close it then it won't be able to find the classes. But it's a shame the error message is not better :smile: 

One thing that there was not any file to comment on:

- :x: Test coverage is lacking. We cannot merge without sufficient test coverage, as we then can't guarantee the feature works.
  - There are no QA tests which load an application with a custom search plugin and test that it works.
  - There are no unit tests in any of the connectors or exporters testing that the custom header plugin is loaded, that it's invoked, etc. 
  - There are no tests which check that an external JAR is properly handled (see for example our [ExternalExporterContainerTest](https://github.com/camunda/camunda/blob/669bdc99296191f5c4a505b6d0cfe851ec05ef51/zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExternalExporterContainerTest.java))

Please address all comments, then re-request a review :)

In the future, I would recommend slicing this into smaller PRs. This would give us a tighter feedback loop, and we would be able to merge and iterate faster on the design as we go along.

For example, one way to split this up:
1. Add the search plugin module
1. Add implementation for the Operate ES connector
2. Add implementation for the Zeebe ES exporter
3. Add implementation for the Tasklist ES connector
4. Add implementation for the Operate OS connector
5. Add implementation for the Zeebe OS exporter
6. Add implementation for the Tasklist OS connector

This would allow us to already have E2E testing for some things early one. It would also allow us to react and maybe do some cleanup/refactoring/deduplication of utility code around the 3rd or 4th PR. It would also make reviews much faster, since each PR would be a lot less to review :)

File: operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java
Comment (igpetrov) [reply operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java]: Changed here and in related OS connector code.

File: operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java
Comment (igpetrov) [reply operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java:224]: I thought to add it in the follow up PR but added in this one

File: tasklist/common/src/main/java/io/camunda/tasklist/es/ElasticsearchConnector.java
Comment (igpetrov) [reply tasklist/common/src/main/java/io/camunda/tasklist/es/ElasticsearchConnector.java]: Changed the message here and in all similar places

File: tasklist/common/src/main/java/io/camunda/tasklist/property/InterceptorPluginProperties.java
Comment (igpetrov) [reply tasklist/common/src/main/java/io/camunda/tasklist/property/InterceptorPluginProperties.java]: Changed as per code style

File: zeebe/exporters/elasticsearch-exporter/pom.xml
Comment (igpetrov) [reply zeebe/exporters/elasticsearch-exporter/pom.xml]: Removed

**Comment by igpetrov** [comment]
Working on tests now 🔁

**Review by npepinpe — CHANGES_REQUESTED** [review]
Thanks!

- :wrench: It would be nice if we could support loading plugins without using a JAR, like with exporters, as it would greatly simplify tests (i.e. not having to create JARs all the time, for example).

File: operate/common/src/main/java/io/camunda/operate/property/InterceptorPluginProperty.java
Comment (igpetrov) [reply operate/common/src/main/java/io/camunda/operate/property/InterceptorPluginProperty.java:44]: Weird 🤔 . Pretty sure I used the same generator as suggested.
I'll double check it

File: operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java
Comment (igpetrov) [reply operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java:244]: I am not sure what else I can add here without writing big text. Maybe I should add a troubleshooting guide here https://github.com/camunda/camunda-docs/pull/4171

File: operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/PluginOperateSearchAbstractIT.java
Comment (igpetrov) [reply operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/PluginOperateSearchAbstractIT.java:34]: Not in this one unfortunately. The problem is this test is running as spring boot:
```
@SpringBootTest(
    classes = {TestApplication.class},
    properties = {
      OperateProperties.PREFIX + ".importer.startLoadingDataOnStartup = false",
      OperateProperties.PREFIX + ".archiver.rolloverEnabled = false",
      "spring.mvc.pathmatch.matching-strategy=ANT_PATH_MATCHER",
      OperateProperties.PREFIX + ".multiTenancy.enabled = false",
      "camunda.operate.zeebeElasticsearch.interceptor-plugins[0].id=" + PLUGIN_ID,
      "camunda.operate.zeebeElasticsearch.interceptor-plugins[0].class-name=" + PLUGIN_CLASS_NAME,
      "camunda.operate.zeebeElasticsearch.interceptor-plugins[0].jar-path="
          + PLUGIN_JAR_BASE_PATH
          + PLUGIN_JAR_FILE_NAME,
      "camunda.operate.elasticsearch.interceptor-plugins[0].id=" + PLUGIN_ID,
      "camunda.operate.elasticsearch.interceptor-plugins[0].class-name=" + PLUGIN_CLASS_NAME,
      "camunda.operate.elasticsearch.interceptor-plugins[0].jar-path="
          + PLUGIN_JAR_BASE_PATH
          + PLUGIN_JAR_FILE_NAME
    })
```

meaning JAR path should be constant. It won't simply compile. Or do you have suggestions?

File: operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/CustomStaticHeaderAddedFromPluginIT.java
Comment (igpetrov) [reply operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/CustomStaticHeaderAddedFromPluginIT.java:63]: In these tests I am testing HTTP requests / interceptors. I cannot do it without loading the context because interceptor is registered in spring configuration.

File: search-plugins/pom.xml
Comment (igpetrov) [reply search-plugins/pom.xml:12]: The plugin SDK will go as a standalone JAR file. I don't see a reason why would we need to fork it from parent.

Comment (npepinpe) [reply search-plugins/pom.xml:12]: The zeebe-parent will become the mono repo parent, and it itself inherits from the camunda-release-parent. The parent only sets dependency management, plugin management, and such things. I'm not sure what that has to do with being a standalone JAR. We already do the same for the `exporter-api`, for example.

Almost all modules (except Optimize, as it was just merged) inherit from the zeebe-parent (directly or indirectly). As you saw in your POM, not doing so means you need to copy a lot of management settings for little benefit.

File: operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java
Comment (igpetrov) [reply operate/common/src/main/java/io/camunda/operate/connect/ElasticsearchConnector.java:244]: I decided to add a section to documentation

File: operate/common/src/main/java/io/camunda/operate/property/InterceptorPluginProperty.java
Comment (igpetrov) [reply operate/common/src/main/java/io/camunda/operate/property/InterceptorPluginProperty.java]: Seems like misclick from my side. Fixed

File: operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/PluginOperateSearchAbstractIT.java
Comment (igpetrov) [reply operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/PluginOperateSearchAbstractIT.java]: Moved these tests to unit-test level

File: operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/CustomStaticHeaderAddedFromPluginIT.java
Comment (igpetrov) [reply operate/qa/integration-tests/src/test/java/io/camunda/operate/elasticsearch/plugin/CustomStaticHeaderAddedFromPluginIT.java]: Moved to unit tests

File: search-plugins/pom.xml
Comment (igpetrov) [reply search-plugins/pom.xml]: Simplified the `pom.xml`

File: search-plugins/src/test/java/io/camunda/zeebe/searchplugin/LoadInterceptorPluginTest.java
Comment (igpetrov) [reply search-plugins/src/test/java/io/camunda/zeebe/searchplugin/LoadInterceptorPluginTest.java]: Removed all together

File: search-plugins/src/test/java/io/camunda/zeebe/searchplugin/impl/TestCustomHeaderInterceptor.java
Comment (igpetrov) [reply search-plugins/src/test/java/io/camunda/zeebe/searchplugin/impl/TestCustomHeaderInterceptor.java]: Removed all together

File: zeebe/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterCustomHeaderPluginIT.java
Comment (igpetrov) [reply zeebe/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterCustomHeaderPluginIT.java]: Did as you suggested: the happy case left as IT; and happy + all others went to unit tests

File: zeebe/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterCustomHeaderPluginIT.java
Comment (igpetrov) [reply zeebe/exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/ElasticsearchExporterCustomHeaderPluginIT.java]: Moved to unit tests

File: zeebe/exporters/opensearch-exporter/src/test/java/io/camunda/zeebe/exporter/opensearch/OpensearchExporterCustomHeaderPluginIT.java
Comment (igpetrov) [reply zeebe/exporters/opensearch-exporter/src/test/java/io/camunda/zeebe/exporter/opensearch/OpensearchExporterCustomHeaderPluginIT.java:1]: Moved to unit tests

File: zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ThreadContextUtil.java
Comment (igpetrov) [reply zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ThreadContextUtil.java:94]: Added javadoc

**Review by npepinpe — APPROVED** [review]
Thanks Igor!

As discussed, I will extract some follow ups to be done, but we can merge this for now to have access to the functionality. I will create issues for them and let you know. I'm pre-approving, but please address any comments before merging (e.g. renaming the test classes so they are suffixed with `Test`).

**Required**
- Implement the interceptor plugin in the new connectors (aka search clients).

**Nice to have**
- Centralize interceptor plugin logic so it's not quadruplicated
  - Not only does this cut down maintenance effort, but it also allows us to test everything in one place (for unit tests).
- Add happy path integration test for Operate ES connector
- Add happy path integration test for Operate OS connector
- Add happy path integration test for Tasklist ES connector
- Add happy path integration test for Tasklist OS connector
- Remove any leftover duplication where it makes sense and brings value (some duplication may be fine, to be evaluated on a case by case basis).

**Comment by igpetrov** [comment]
Hi @npepinpe, thank you for pre-approving!

I will start working on the follow-ups right away in the next PR.

**Comment by npepinpe** [comment]
Please wait until I've created issues. We can also implement them in multiple PRs.

**Comment by npepinpe** [comment]
So I doubled check, it seems the new connector stuff is not yet used, so we can ignore it for now. I will still create a follow up for whoever picks that project up that they need to integrate it anyway.

**Comment by igpetrov** [comment]
I guess, I likely will be the person who will work on follow-ups.

> it seems the new connector stuff is not yet used, so we can ignore it for now. I will still create a follow up for whoever picks that project up that they need to integrate it anyway

Makes sense.

> Move the search-plugin module under camunda-search, and name it camunda-search-client-plugin.

Makes sense too. Maybe I will also can sync on it with @romansmirnov about what was his vision on both things. WDYT?

File: search-plugins/src/main/java/io/camunda/plugin/search/header/CustomHeader.java
Comment (igpetrov) [reply search-plugins/src/main/java/io/camunda/plugin/search/header/CustomHeader.java:17]: Done

File: operate/common/src/test/java/io/camunda/operate/connect/OperateESPluginLoading.java
Comment (igpetrov) [reply operate/common/src/test/java/io/camunda/operate/connect/OperateESPluginLoading.java]: Done

**Comment by npepinpe** [comment]
I'll check when Roman is back from FTO. If it's in a while, I think me and someone else can make a decision.

**Comment by npepinpe** [comment]
FYI - https://github.com/camunda/camunda/issues/21559. I will write further sub issues. We can sync after to make sure we're on the same page.

**Pull request merged**
